### PR TITLE
[NFCI][SYCL] Eliminate UB in ANSI alias violation accessing ByteArray's

### PR DESCRIPTION
**Pull request by aelovikov-intel**
As part of it provide a nicer (IMO) interface to it. Also, asserts are now moved into the underlying utility functions making it easier to read the core logic up the call stack.

### CODE REVIEW

File: sycl/source/detail/device_image_impl.hpp [hunk 1]
```diff
@@ -260,39 +260,33 @@ class device_image_impl {
...
           // should not have a descriptor. These padding elements all have max
           // ID value.
-          if (It[0] != std::numeric_limits<std::uint32_t>::max()) {
+          if (Id != std::numeric_limits<std::uint32_t>::max()) {
             // The map is not locked here because updateSpecConstSymMap() is
             // only supposed to be called from c'tor.
             MSpecConstSymMap[std::string{SCName}].push_back(
-                SpecConstDescT{/*ID*/ It[0], /*CompositeOffset*/ It[1],
-                               /*Size*/ It[2], BlobOffset});
+                SpecConstDescT{Id, CompositeOffset, Size, BlobOffset});
```
Comment (aelovikov-intel) [inline sycl/source/detail/device_image_impl.hpp:286]: Could be 

```
MSpecConstSymMap[std::string{SCName}].emplace_back(Id, CompositeOffset, Size, BlobOffset);
```

but should be done in a different PR (if at all).

File: sycl/source/detail/device_binary_image.hpp [hunk 2]
```diff
@@ -27,9 +28,35 @@ class ByteArray {
...
+
+  bool empty() const { return Size == 0; }
+
 private:
+  template <typename T> T consumeOneElem() {
+    assert(sizeof(T) <= Size && "Out of bounds!");
+    T Val;
+    std::memcpy(&Val, Ptr, sizeof(T));
+    Ptr += sizeof(T);
+    Size -= sizeof(T);
```
Comment (steffenlarsen) [inline sycl/source/detail/device_binary_image.hpp:54]: Nit, but you could call `dropBytes(sizeof(T))` here instead. Should make maintainability a little easier, in case something else comes along that needs to be mutated with every move.

**Review by steffenlarsen — APPROVED** [review]
I agree, this looks great! Just one minor comment, but I don't think it needs to block this.

**Comment by aelovikov-intel** [comment]
@intel/llvm-gatekeepers , this PR is ready.
